README.md
script.duration
===============

Displays movies duration in hours and minutes instead of only minutes while navigating in Kodi 17 movies library. Kodi skin has to be tweaked, see exemple below.

Exemple: __108 minutes__ is now displayed __1h48__.

When launched the script provides those properties :

* `Window(Home).Property(Durations.HoursMinutes)`
* `Window(Home).Property(Durations.Hours)`
* `Window(Home).Property(Durations.Minutes)`
* `Window(Home).Property(Durations.DBID)`

### Exemple of integration in Estuary skin

2 files need to be modified as follow :

#### Variables.xml

Add this new variable at the end of the file (before the `</include>`) :
```
<variable name="ItemDuration">
    <value condition="System.HasAddon(script.duration)+
        [Window.IsVisible(Videos) | Window.IsVisible(Movieinformation)] +
        !String.IsEmpty(Window(Home).Property(Durations.Hours)) +
        !String.IsEqual(Window(Home).Property(Durations.Hours),0) +
        !String.IsEmpty(Window(Home).Property(Durations.DBID)) +
        String.IsEqual(Window(Home).Property(Durations.DBID),ListItem.DBID)">$INFO[Window(Home).Property(Durations.HoursMinutes)]</value>
    <value>$INFO[ListItem.Duration] min</value>
</variable>
```

#### Includes.xml

Replace `<label>$INFO[$PARAM[infolabel_prefix]ListItem.Duration]</label>` with `<label>$VAR[ItemDuration]</label>`

______________________


changelog.txt
[B]0.3.8[/B] (2018-04-26)
- Minor improvements and optimizations

[B]0.3.7[/B] (2018-04-19)
- Minor improvements and optimizations

[B]0.3.5[/B] (2017-07-14)
- Reworked script, changed/removed deprecated stuff
- New property names (see description for details)

[B]0.3.4[/B] (2017-07-11)
- ListItem.DBID is now also set - for verification purposes (Duration.DBID)

[B]0.3.3[/B] (2016-08-03)
- Minor cleanup

[B]0.3.2[/B] (2016-07-23)
- axbmcuser MOD: Fix for KODI 17 (Krypton): "videolibrary" renamed to "videos" [default.py]

[B]0.3.1[/B]
- axbmcuser MOD: Minor improvements and optimizations

[B]0.3[/B]
- axbmcuser MOD: Script now outputs H:MM

[B]0.2[/B]
- Correction of crash when duration value isn't properly written in database

[B]0.1[/B]
- Initial release


default.py
# -*- coding: UTF-8 -*-

import xbmc, xbmcaddon, xbmcgui

# Script constants
__addon__      = xbmcaddon.Addon()
__addonid__    = __addon__.getAddonInfo('id')
__version__    = __addon__.getAddonInfo('version')
__language__   = __addon__.getLocalizedString
__cwd__        = __addon__.getAddonInfo('path')

def log(txt):
    if isinstance (txt,str):
        txt = txt.decode("utf-8")
    message = u'%s: %s' % (__addonid__, txt)
    xbmc.log(msg=message.encode("utf-8"), level=xbmc.LOGDEBUG)

def in_hours_and_min(minutes_string):
    try:
        full_minutes = int(minutes_string)
        minutes = full_minutes % 60
        hours   = full_minutes // 60
        return str(hours) + 'h' + str(minutes).zfill(2)
    except:
        return '-h--'

class Main:
    def __init__( self ):
        log("version %s started" % __version__)
        self._init_vars()
        self._parse_argv()
        # don't run if already in backend
        if xbmc.getCondVisibility("IsEmpty(Window(videolibrary).Property(duration_backend_running))"):
            if self.backend:
                # run in backend if parameter was set
                xbmc.executebuiltin("SetProperty(duration_backend_running,true,videolibrary)")
                self.run_backend()
            if self.duration:
                self.display_duration()

    def _init_vars(self):
        #winid = xbmcgui.getCurrentWindowId()
        #log('window ID: ' + str(winid))
        self.window = xbmcgui.Window(10025) # MyVideoNav.xml (videolibrary)
        self.dialogwindow = xbmcgui.Window(12003) # DialogVideoInfo.xml (movieinformation)

    def _parse_argv( self ):
        try:
            params = dict( arg.split( '=' ) for arg in sys.argv[ 1 ].split( '&' ) )
        except:
            params = {}
        log("params: %s" % params)
        self.duration  = params.get( 'duration', False )
        self.backend = params.get( 'backend', False )

    def run_backend(self):
        self._stop = False
        self.previousitem = ""
        while not self._stop:
            if not xbmc.getCondVisibility("Container.Scrolling"):
                self.selecteditem = xbmc.getInfoLabel("ListItem.DBID")
                if (self.selecteditem != self.previousitem):
                    self.previousitem = self.selecteditem
                    if xbmc.getInfoLabel("ListItem.DBID") > -1 and not xbmc.getCondVisibility("ListItem.IsFolder"):
                        self.duration = xbmc.getInfoLabel("ListItem.Duration")
                        self.display_duration()
            xbmc.sleep(200)
            if not xbmc.getCondVisibility("Window.IsVisible(videolibrary)"):
                log('backend stopped.')
                xbmc.executebuiltin('ClearProperty(duration_backend_running,videolibrary)')
                self._stop = True

    def display_duration(self):
        log('Converts : '+self.duration+' min.')
        readable_duration = in_hours_and_min(self.duration)
        self.window.setProperty('Duration', readable_duration)
        self.dialogwindow.setProperty('Duration', readable_duration)

if (__name__ == "__main__"):
    Main()
log('script finished.')


service.py
# -*- coding: UTF-8 -*-

import xbmc, xbmcaddon, xbmcgui

# Script constants
__addon__      = xbmcaddon.Addon()
__addonid__    = __addon__.getAddonInfo('id')
__version__    = __addon__.getAddonInfo('version')
__language__   = __addon__.getLocalizedString
__cwd__        = __addon__.getAddonInfo('path')

def log(txt):
    if isinstance (txt,str):
        txt = txt.decode("utf-8")
    message = u'%s: %s' % (__addonid__, txt)
    xbmc.log(msg=message.encode("utf-8"), level=xbmc.LOGDEBUG)

def get_hours_and_minutes(minutes_string):
    try:
        full_minutes = int(minutes_string)
        minutes = full_minutes % 60
        hours   = full_minutes // 60
        return str(hours) + 'h' + str(minutes).zfill(2)
    except:
        return ''

def get_hours_only(minutes_string):
    try:
        full_minutes = int(minutes_string)
        hours   = full_minutes // 60
        return str(hours)
    except:
        return ''

def get_minutes_only(minutes_string):
    try:
        full_minutes = int(minutes_string)
        minutes = full_minutes % 60
        return str(minutes).zfill(2)
    except:
        return ''

class Main:
    def __init__( self ):
        log("version %s started" % __version__)
        self.previousitem = ""
        self.monitor = xbmc.Monitor()
        self.run_service()

    def run_service(self):
        while not self.monitor.abortRequested():
            if self.monitor.waitForAbort(0.2):
                break
            if ((xbmc.getCondVisibility("Window.IsActive(Videos)") and xbmc.getCondVisibility("Window.Is(Videos)")) or (xbmc.getCondVisibility("Window.IsActive(MovieInformation)") and xbmc.getCondVisibility("Window.Is(MovieInformation)"))) and not xbmc.getCondVisibility("Container.Scrolling"):
                self.selecteditem = xbmc.getInfoLabel("ListItem.DBID")
                if (self.selecteditem and self.selecteditem != self.previousitem):
                    self.previousitem = self.selecteditem
                    
                    if (xbmc.getInfoLabel("ListItem.DBID") and xbmc.getInfoLabel("ListItem.DBTYPE") and (xbmc.getInfoLabel("ListItem.DBTYPE") == 'movie' or xbmc.getInfoLabel("ListItem.DBTYPE") == 'episode') and not xbmc.getCondVisibility("ListItem.IsFolder")) and xbmc.getInfoLabel("ListItem.Duration"):
                        try:
                            tmpval = float(int(xbmc.getInfoLabel("ListItem.Duration")))
                            if (tmpval > 0) :
                                self.duration = xbmc.getInfoLabel("ListItem.Duration")
                                self.dbid = xbmc.getInfoLabel("ListItem.DBID")
                                self.set_duration_values()
                        except:
                            pass
            else:
                my_container_id = xbmc.getInfoLabel("Window(Home).Property(Durations.WidgetContainerId)")
                my_container_window = xbmc.getInfoLabel("Window(Home).Property(Durations.WidgetContainerWindowName)")
                
                if (my_container_id and my_container_window and (xbmc.getCondVisibility("Control.HasFocus("+my_container_id+")") and xbmc.getCondVisibility("Window.IsActive("+my_container_window+")") and xbmc.getCondVisibility("Window.Is("+my_container_window+")")) and not xbmc.getCondVisibility("Window.IsActive(Videos)") and not xbmc.getCondVisibility("Window.IsActive(MovieInformation)")) and not xbmc.getCondVisibility("Container("+my_container_id+").Scrolling"):
                    self.selecteditem = xbmc.getInfoLabel("Container("+my_container_id+").ListItem.DBID")
                    if (self.selecteditem and self.selecteditem != self.previousitem):
                        self.previousitem = self.selecteditem
                        if (xbmc.getInfoLabel("Container("+my_container_id+").ListItem.DBID") and xbmc.getInfoLabel("Container("+my_container_id+").ListItem.DBTYPE") and (xbmc.getInfoLabel("Container("+my_container_id+").ListItem.DBTYPE") == 'movie' or xbmc.getInfoLabel("Container("+my_container_id+").ListItem.DBTYPE") == 'episode') and not xbmc.getCondVisibility("Container("+my_container_id+").ListItem.IsFolder")) and xbmc.getInfoLabel("Container("+my_container_id+").ListItem.Duration"):
                            try:
                                tmpval = float(int(xbmc.getInfoLabel("Container("+my_container_id+").ListItem.Duration")))
                                if (tmpval > 0) :
                                    self.duration = xbmc.getInfoLabel("Container("+my_container_id+").ListItem.Duration")
                                    self.dbid = xbmc.getInfoLabel("Container("+my_container_id+").ListItem.DBID")
                                    self.set_duration_values()
                            except:
                                pass
    #run_service end

    def set_duration_values(self):
        log('Converts: '+self.duration+' min')
        # HoursMinutes
        hours_and_minutes = get_hours_and_minutes(self.duration)
        xbmc.executebuiltin('SetProperty(Durations.HoursMinutes,"'+hours_and_minutes+'",home)')
        # Hours
        hours_only = get_hours_only(self.duration)
        xbmc.executebuiltin('SetProperty(Durations.Hours,"'+hours_only+'",home)')
        # Minutes
        minutes_only = get_minutes_only(self.duration)
        xbmc.executebuiltin('SetProperty(Durations.Minutes,"'+minutes_only+'",home)')
        # DBID
        xbmc.executebuiltin('SetProperty(Durations.DBID,"'+self.dbid+'",home)')
        # InputDurationMinutes
        xbmc.executebuiltin('SetProperty(Durations.InputDurationMinutes,"'+self.duration+'",home)')

if (__name__ == "__main__"):
    Main()
log('script finished.')
